javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/PatchByteCodeVersionTransformer.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.tooling.instrumentation.indy;
+
+import static org.objectweb.asm.ClassWriter.COMPUTE_FRAMES;
+
+import java.security.ProtectionDomain;
+import net.bytebuddy.ClassFileVersion;
+import net.bytebuddy.agent.builder.AgentBuilder;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.asm.AsmVisitorWrapper;
+import net.bytebuddy.description.field.FieldDescription;
+import net.bytebuddy.description.field.FieldList;
+import net.bytebuddy.description.method.MethodList;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.dynamic.DynamicType;
+import net.bytebuddy.implementation.Implementation;
+import net.bytebuddy.pool.TypePool;
+import net.bytebuddy.utility.JavaModule;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.commons.JSRInlinerAdapter;
+
+/**
+ * Patches the class file version to 51 (Java 7) in order to support injecting {@code INVOKEDYNAMIC}
+ * instructions via {@link Advice.WithCustomMapping#bootstrap} which is important for indy plugins.
+ */
+public class PatchByteCodeVersionTransformer implements AgentBuilder.Transformer {
+
+  private static boolean isAtLeastJava7(TypeDescription typeDescription) {
+    ClassFileVersion classFileVersion = typeDescription.getClassFileVersion();
+    return classFileVersion != null && classFileVersion.getJavaVersion() >= 7;
+  }
+
+  @Override
+  public DynamicType.Builder<?> transform(
+      DynamicType.Builder<?> builder,
+      TypeDescription typeDescription,
+      ClassLoader classLoader,
+      JavaModule javaModule,
+      ProtectionDomain protectionDomain) {
+
+    if (isAtLeastJava7(typeDescription)) {
+      // we can avoid the expensive (and somewhat dangerous) stack frame re-computation if stack
+      // frames are already present in the bytecode, which also allows eagerly loading types that
+      // might be present in the method body, but not yet loaded by the JVM.
```

Comment:
> which also allows eagerly loading types that might be present in the method body, but not yet loaded by the JVM

I think this is an important comment, can you expand a bit on it, including any ramifications of eagerly loading types that may not yet be loaded by the JVM?

Reply:
This is the PR that provides the initial context: https://github.com/elastic/apm-agent-java/pull/1758/files

We had a few reports of deadlocks that were caused by the expensive stack frame re-computation: https://github.com/elastic/apm-agent-java/issues/1839 and https://github.com/elastic/apm-agent-java/issues/1884, in particular when the instrumentations is applied dynamically at runtime (for example when a listener instance is passed as an argument and the actual type of the listener can't be guessed in advance by the agent without scanning every class).

Those issues have been solved (or at least their symptoms disappeared) once we did two things:
- bypass when it's not needed (which at the time required a change in bytebuddy)
- stop relying on classloader for resolving types (as re-computing stack frames requires to know a bit about the types that are involved in the method body) which might alter the usual class-loading order in the application.

If you think that's clear enough, add a :+1: on this comment and I will change the comment to elaborate a bit more.

Speaking of the latter, I have completely forgotten to change this behavior, so I will need to change that if needed.

Reply:
The agent is already using the `POOL_ONLY` strategy to resolve types ([link](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/acbab58a4696169802595e75e738572685abad0c/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/AgentInstaller.java#L146)), thus no change is required here to avoid classloading side-effects.

Reply:
I don't get what you mean by
> which also allows eagerly loading types that might be present in the method body, but not yet loaded by the JVM.

Is this about using `AgentBuilder.DescriptionStrategy.Default.HYBRID`? If so I think this comment is not relevant, agent just can not safely load classes during transform.

Reply:
Yes, you are correct here @laurit , in the original context this comment was written the `HYBRID` strategy was initially used. I'll simplify it to this (that will also resolve this conversation when I push).
```
we can avoid the expensive stack frame re-computation if stack frames are already present in the bytecode.
```